### PR TITLE
[RFR] Facets

### PR DESCRIPTION
**Pull request by djhi**
[Trello card](https://trello.com/c/dagbcExp)

- [x] Added `isFacet` property to fields
- [x] Added and populated `publishedFacet` collection
- [x] Public UI for facets
- [x] Tests

### CODE REVIEW

File: src/api/controller/api/publish.js [hunk 1]
```diff
@@ -103,6 +112,8 @@ export const doPublish = async (ctx) => {
         transformDocumentAndKeepUri,
     );
     await ctx.publishCharacteristics(ctx, datasetCoverFields, count);
+    console.log({ facetFields });
```
Comment (ThieryMichel) [inline src/api/controller/api/publish.js]: debug spotted

File: src/api/models/publishedFacet.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    collection.insertBatch = documents => chunk(documents, 100).map(data => collection.insertMany(data));
+
+    collection.insertFacet = (field, values) =>
+        collection.insertBatch(values.map(value => ({
+            field,
+            value,
+        })));
+
+    collection.findLimitFromSkip = (limit, skip, filter) =>
```
Comment (ThieryMichel) [inline src/api/models/publishedFacet.js:14]: refactor to return both count and data
```js
const cursor = collection.find(...).sort(...)

return {
    count: await cursor.count(),
    data: await cursor.skip(...).limit(...).toArray(),
};
```

File: src/api/models/publishedFacet.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        collection
+            .find(filter)
+            .skip(skip)
+            .limit(limit)
+            .sort({ value: 1 })
```
Comment (ThieryMichel) [inline src/api/models/publishedFacet.js:19]: sort before skip and limit otherwise you sort a subset of the data

**Review by ThieryMichel — CHANGES_REQUESTED** [review]

File: src/api/models/publishedFacet.js (hunk 3 above)
Comment (djhi) [reply src/api/models/publishedFacet.js:19]: I'm quite sure that's not true as the query is only executed when yielding `toArray`

File: src/api/models/publishedFacet.js (hunk 2 above)
Comment (djhi) [reply src/api/models/publishedFacet.js:14]: Will do for all models in a separate PR

**Review by ThieryMichel — APPROVED** [review]

**Pull request merged**